### PR TITLE
Implementation of recommendation 3.2

### LINKED ISSUE #66: Recommendation 3.2: The declaration part of a class-like code structures (class/interface/e.t.c) should be in the proper order
Notes:

- There should be no blank lines between properties without comments;
- Properties with comments/Kdoc should be separated by a newline before the comment/Kdoc

The declaration part of a class or interface should be in the following order:

* compile-time constants (for objects)
* Properties
* late-init class properties
* init-blocks
* constructors
* methods or nested classes. Put nested classes next to the code that uses those classes. If the classes are intended to be used externally and aren't referenced inside the class, put them in the end, after the companion object.
* companion object
* *Exception*： All variants of a private val logger should be placed in the beginning of the class (private val log/LOG/logger/e.t.c)

### PR DESCRIPTION
**Pull request by petertrr**
This closes #66 

### What's done:
* Introduced ASTNodeUtils#checkOrder, added to FileStructureRule
* Refactored `lintMethod` in `TestUtils` to make assertion errors more readable
* Introduced ClassLikeStructuresOrderRule.kt
* Added tests on checkers and fixers
* Updated available-rules.md
* Updated rules-config.json

### CODE REVIEW

**Comment by akuleshov7** [comment]
you confused me a little bit - it is a Rule 3.2, not a recommendation :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import org.cqfn.diktat.ruleset.utils.leaveOnlyNNewLines
+import org.cqfn.diktat.ruleset.utils.moveChildBefore
+import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import org.jetbrains.kotlin.com.intellij.psi.tree.TokenSet
+
+/**
+ * Rule that checks order of declarations inside classes, interfaces and objects.
+ */
+class ClassLikeStructuresOrderRule : Rule("class-like-structures") {
+    private val loggerPropertyRegex = "(log|LOG|logger)".toRegex()
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:42]: please move it to some common place, that will be usefull

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        if (node.elementType == CLASS_BODY) {
+            checkDeclarationsOrderInClass(node)
+        } else if (node.elementType == PROPERTY) {
+            checkNewLinesBeforeProperty(node)
+        }
+    }
+
+    private fun checkDeclarationsOrderInClass(node: ASTNode) {
+        val allProperties = node.getChildren(TokenSet.create(PROPERTY))
+        val constProperties = allProperties.filter { it.findLeafWithSpecificType(CONST_KEYWORD) != null }
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:65]: consts can be only in object companion, no?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        val lateInitProperties = allProperties.filter { it.findLeafWithSpecificType(LATEINIT_KEYWORD) != null }
+        val loggers = allProperties.filter {
+            it.findLeafWithSpecificType(PRIVATE_KEYWORD) != null && it.getIdentifierName()!!.text.contains(loggerPropertyRegex)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:68]: loggers can be not only private, also they are commonly placed to companion object

**Comment by petertrr** [comment]
> you confused me a little bit - it is a Rule 3.2, not a recommendation :)

Why, no: 
`Rule 3.2 A source file contains code blocks in the following order: copyright, package name, imports, top-level classes.`
`Recommendation 3.2: The declaration part of a class-like code structures (class/interface/e.t.c) should be in the following order`

Also, you have named issue #66 `recommendation 3.2` :)

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt (hunk 2 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:65]: Yes. `companion object`'s body is also a node of type `CLASS_BODY`, so the list `constProperties` will be not empty only for `companion object`s.

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt (hunk 3 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:68]: To quote our code style guide:
`Exception： All variants of a private val logger should be placed in the beginning of the class (private val log/LOG/logger/e.t.c)`.

It is quite vaguely stated. Should I change it both in code and in guide?

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        val (usedClasses, unusedClasses) = node.getChildren(TokenSet.create(CLASS)).partition { classNode ->
+            classNode.getIdentifierName()?.let { identifierNode ->
+                node.parent(FILE)!!.findAllNodesWithSpecificType(REFERENCE_EXPRESSION).any { ref ->
+                    ref.parent({ it == classNode }) == null && ref.text.contains(identifierNode.text)
+                }
+            } ?: false
+        }
+        val companion = node.getChildren(TokenSet.create(OBJECT_DECLARATION))
+                .find { it.findChildByType(MODIFIER_LIST)?.findLeafWithSpecificType(COMPANION_KEYWORD) != null }
+        val blocks = Blocks(AllProperties(loggers.toMutableList(), constProperties.toMutableList(), properties.toMutableList(), lateInitProperties.toMutableList()),
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:83]: move `toMutableList` to map function

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt [hunk 5]
```diff
@@ -257,3 +266,19 @@ fun ASTNode.areChildrenBeforeGroup(children: List<ASTNode>, group: List<ASTNode>
     return children.map { getChildren(null).indexOf(it) }.max()!! < group.map { getChildren(null).indexOf(it) }.min()!!
 }
 
+fun List<ASTNode>.checkOrder(getSiblingBlocks: ASTNode.() -> Pair<ASTNode?, ASTNode>,
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/utils/AstNodeUtils.kt:269]: I don't like this method :( 

did you want to return anything by this method? 
if you are fixing, not checking - change the name of this method

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+    override fun visit(node: ASTNode,
+                       autoCorrect: Boolean,
+                       params: KtLint.Params,
+                       emit: (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit) {
+        configRules = params.getDiktatConfigRules()
+        emitWarn = emit
+        isFixMode = autoCorrect
+
+        if (node.elementType == CLASS_BODY) {
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:56]: Please create an issue on me to extend code style for Files

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    private fun checkNewLinesBeforeProperty(node: ASTNode) {
+        val previousProperty = node.prevSibling { it.elementType == PROPERTY }
+
+        if (previousProperty != null) {
+            val commentOnThis = node.findChildByType(TokenSet.create(KDOC, EOL_COMMENT, BLOCK_COMMENT))
+            val whiteSpaceBefore = previousProperty.nextSibling { it.elementType == WHITE_SPACE }!!
+            val nRequiredNewLines = if (commentOnThis == null) 1 else 2
+            if (whiteSpaceBefore.text.count { it == '\n' } != nRequiredNewLines)
+                BLANK_LINE_BETWEEN_PROPERTIES.warnAndFix(configRules, emitWarn, isFixMode, node.getIdentifierName()!!.text, node.startOffset) {
+                    whiteSpaceBefore.leaveOnlyNNewLines(nRequiredNewLines)
```
Comment (akuleshov7) [inline diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:109]: this naming will not fit our code style

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt (hunk 6 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:56]: Created #80

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt (hunk 1 above)
Comment (petertrr) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:42]: Yes, will move to StringUitls.kt

File: diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt (hunk 3 above)
Comment (akuleshov7) [reply diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/ClassLikeStructuresOrderRule.kt:68]: ah, I got it! Please fix code style to: (private) val

**Pull request merged**